slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala

```diff
@@ -514,7 +514,7 @@ trait JdbcActionComponent extends SqlActionComponent { self: JdbcProfile =>
       session.withPreparedStatement(sql)(f)
 
     class SingleInsertAction(a: compiled.Artifacts, value: U) extends SimpleJdbcProfileAction[SingleInsertResult]("SingleInsertAction", Vector(a.sql)) {
-      def run(ctx: Backend#Context, sql: Vector[String]) = preparedInsert(a.sql, ctx.session) { st =>
+      def run(ctx: Backend#Context, sql: Vector[String]) = preparedInsert(sql.head, ctx.session) { st =>
```

Comment:
Hi, I don't know the codebase well enough to understand this change. Can you explain what's happening? What is `a` and what gets passed as `sql` to `run`? Can you ELI5 why the old way is fails and the new way fixes it? Thanks.

Reply:
Hi, thanks for looking into this. I don't know the codebase well, either, but I'll try to explain. 🙂 

`SingleInsertAction` and `MultiInsertAction` extend `SimpleJdbcProfileAction`. The class has a complicated structure which seems to have been introduced to support statement overrides: https://github.com/slick/slick/blob/98ded3a777372d0783dbb42233b08ff51f6da3a7/slick/src/main/scala/slick/jdbc/JdbcActionComponent.scala#L25-L32
It seems that `overrideStatements` depends on the fact that `run(ctx, sql)` actually uses `sql` to prepare a statement, and not the argument `a` passed to the constructor of `SingleInsertAction` or `MultiInsertAction`. 

Now, when the `overrideStatements` isn't called, `a.sql` ends up being passed to `run(ctx, sql)` anyway, because:
- `Vector(a.sql)` is passed to `SimpleJdbcProfileAction` as a constructor argument.
- `SimpleJdbcProfileAction#run(ctx)` calls `run(ctx, sql)` where `sql` is the value of this constructor argument.
- `run(ctx, sql)` is only ever called inside `SimpleJdbcProfileAction`. Its visibility could probably be changed to `protected` and it should still work.

Let me know if this answers your questions.

Reply:
If you think this would help, I can try to refactor these classes a bit.

Reply:
Makes sense, thanks for the explanation!